libsweep/include/queue.h

```diff
@@ -0,0 +1,68 @@
+#ifndef SWEEP_QUEUE_62C8F42E8DD5_H
+#define SWEEP_QUEUE_62C8F42E8DD5_H
+
+/*
+ * Thread-safe queue.
+ * Implementation detail; not exported.
+ */
+
+#include <stdint.h>
+
+#include <atomic>
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+#include <utility>
+
+namespace sweep {
+namespace queue {
+
+template <typename T> class queue {
+public:
+  queue(int32_t max) : max_size(max), the_queue(), the_mutex(), the_cond_var() {}
+
+  // Empty the queue
+  void flush() {
+    std::unique_lock<std::mutex> lock(the_mutex);
+    while (!the_queue.empty()) {
+      the_queue.pop();
+    }
+  }
+
+  // Add an element to the queue.
+  void enqueue(T v) {
+    std::lock_guard<std::mutex> lock(the_mutex);
+
+    // if necessary, remove the oldest scan to make room for new
+    if (static_cast<int32_t>(the_queue.size()) >= max_size)
+      the_queue.pop();
+
+    the_queue.push(v);
+    the_cond_var.notify_one();
+  }
+
+  // If the queue is empty, wait till an element is avaiable.
+  T dequeue() {
+    std::unique_lock<std::mutex> lock(the_mutex);
+    // wait until queue is not empty
+    while (the_queue.empty()) {
+      // the_cond_var could wake up the thread spontaneously, even if the queue is still empty...
+      // so put this wakeup inside a while loop, such that the empty check is performed whenever it wakes up
+      the_cond_var.wait(lock); // release lock as long as the wait and reaquire it afterwards.
+    }
```

Comment:
There is an overload for `.wait` that takes a predicate which could shorten this a bit:
http://en.cppreference.com/w/cpp/thread/condition_variable/wait

    the_cond_var.wait(lock, [this]{ return !this->the_queue.empty(); } );

Whether this would make the code actually more readable is up for debate